Experiments public contract

wzieba opened the pull request:
### Description

This PR introduces improvements to the public contract:
- extracts an interface: `VariationsRepository`, so the clients can easily double this in tests + KDoc comments
- introduces a behavior change: refreshing cache on SDK initialization, as I believe this is something that every client would do on initialization anyway
- simplifies `refresh` API (no need for two methods)
- adds `README.md` with an explanation about how to use the SDK

### Testing

The only behavior change has been covered  with automated tests. It's still possible to smoke test using instructions from #241 , if necessary.

wzieba committed:
Enable Kotlin's strict API mode and update visibility modifiers

wzieba committed:
Extract an `VariationsRepository` interface

Because this is what ExPlat practically is - a repository for `Variation`.

Extracting an interface makes it easier for consumers to introduce a test double for this

wzieba committed:
Add KDoc documentation to `VariationsRepository` interface

wzieba committed:
Replace `refreshIfNeeded` and `forceRefresh` with a single `refresh` method

This was it's more readable IMO and is similar to how other systems work (especially in CMD) - the operation is requested and can perform the action or not depending on circumstances. But it also has a `-f` parameter.

wzieba committed:
Introduce API verification

So we can be aware of any changes to the public contract.

wzieba committed:
Add a comment about `VariationRepository` being a singleton

wzieba committed:
Fix typo in pipeline.yml

wzieba committed:
Make the `Experiment` a value class and not interface

I don't think `interface` is very useful here. Value class is just fine and makes the API cleaner

wzieba committed:
Add README.md with instructions on how to use the SDK

wzieba committed:
Fetch new variations if cache is stale on ExPlat initialization

We want SDK to provide easy API, and not bother clients with necessary requests (like calling `refresh` manually on every initialization)

wzieba commented on experimentation/src/main/java/com/automattic/android/experimentation/ExPlat.kt:
```diff
@@ -5,50 +5,38 @@ import com.automattic.android.experimentation.ExPlat.RefreshStrategy.IF_STALE
+public class ExPlat internal constructor(
     private val platform: String,
-    private val experiments: Set<Experiment>,
+    experiments: Set<Experiment>,
```
I wonder: is there a value in providing experiments in the constructor and then checking for its existence in `getVariation`? I don't see good benefits 🤔

ParaskP7 commented on .buildkite/pipeline.yml:
```diff
@@ -30,10 +30,18 @@ steps:
+    plugins: [$CI_TOOLKIT]
+    command: |
+      cp gradle.properties-example gradle.properties
+      ./gradlew :experimentation:apiCheck
```
Praise (❤️): Awesome, love it, thanks for adding the `binary-compatibility-validator` plugin into this repo, let's start using this more and more within this project and other repos as well! 🥇

ParaskP7 commented on experimentation/README.md:
```diff
@@ -0,0 +1,63 @@
+# Experimentation
```
Praise (❤️): Thanks for creating a `README.md` to help with the API usage!

ParaskP7 commented on experimentation/src/main/java/com/automattic/android/experimentation/ExPlat.kt:
```diff
@@ -5,50 +5,38 @@ import com.automattic.android.experimentation.ExPlat.RefreshStrategy.IF_STALE
-    fun configure(anonId: String? = null) {
+    init {
+        coroutineScope.launch {
+            refresh()
```
Question (❓): I think this is a good move, but how about when users would like to also use `configure(...)` to setup ExPlat with an anonymous id, they will then need to trigger `refresh()` manually once more, right, or, how will they better use the API then? 🤔

wzieba replied:
True, this was not the best idea. I learned that `anonId` is actually a mandatory information ( p1725007206945009-slack-C7HH3V5AS) and I proposed changes to this initialization methods in #245

ParaskP7 replied:
Thanks for all this work @wzieba , I followed all the conversation on anonymous id needing to become mandatory (and all), plus did a quick review on #245 , the changes makes sense to me (also, thanks @iangmaia for the review)! 🚀

ParaskP7 commented on experimentation/src/main/java/com/automattic/android/experimentation/VariationsRepository.kt:
```diff
@@ -0,0 +1,85 @@
+ *
+ * Should be a Singleton, as it keeps track of the "active" [Variation]s. See [getVariation] for more details.
+ */
+public interface VariationsRepository {
```
Praise (❤️): Love it! 🥇

ParaskP7 reviewed (commented):
And with that final review, I think I am done with all the review @wzieba , kudos on this work, it will help our product teams setup and start using [ExPlat](https://experiments.a8c.com/experiments) without much overhead, and mostly, without `FluxC`! 🎉 🥇 🚀 

Let's start resolving the comments one-by-one and merge everything to `trunk` when ready, just let me know how to help you from here! 💯

wzieba commented on experimentation/src/main/java/com/automattic/android/experimentation/VariationsRepository.kt:
```diff
@@ -0,0 +1,85 @@
+            isDebug: Boolean,
+            cacheDir: File,
+        ): ExPlat {
+            val restClient = ExperimentRestClient(dispatcher = dispatcher)
+            val cache = FileBasedCache(cacheDir, dispatcher = dispatcher, scope = coroutineScope)
+            val assignmentsRepository = AssignmentsRepository(restClient, cache)
+            val assignmentsValidator = AssignmentsValidator(SystemClock())
+            return ExPlat(
+                platform = platform,
+                experiments = experiments,
+                logger = logger,
+                coroutineScope = coroutineScope,
+                isDebug = isDebug,
+                assignmentsValidator = assignmentsValidator,
+                repository = assignmentsRepository,
+            )
```
- [ ] https://github.com/Automattic/Automattic-Tracks-Android/pull/241#discussion_r1736393960
- [ ] https://github.com/Automattic/Automattic-Tracks-Android/pull/242#discussion_r1736610369

wzieba committed:
Merge branch 'experimentation_refactor' into experiments_public_contract

wzieba committed:
Adjust `VariationsRepository` signature

wzieba committed:
Save `anon id` in local cache

This commit adds a new class: `CacheDto` that saves previously saved data with addition to `anonymousId`. This additional field will be used for cache validation.

wzieba committed:
Improve SDK initialization

The `initialize` method now takes a mandatory `anonymousId` parameter and is responsible for setting the anonymous identifier and triggering the initial assignment refresh and cache validation if `anonymousId` is different

wzieba committed:
Simplify public contract: remove `VariationsRepository#clear`

I came to a conclusion, that `refresh` is not needed for the client: SDK should on its own manage the state of the cache and not delegate additional responsibility to SDK consumer.

The cache is refreshed whenever needed during initialization

wzieba committed:
Refactor: remove `RefreshStrategy`

As SDK doesn't offer refreshing anymore, we don't need to use internal-only `RefreshStrategy`. Refactor seems successful as unit tests are still passing without any changes

wzieba committed:
Guard against not initialized sdk

If ExPlat was not initialized, log or throw an exception for quicker debugging

wzieba committed:
Refactor guarding against empty experiments

wzieba committed:
Adjust the sample app to the new API

wzieba committed:
Make `Assignments` internal

It doesn't have to be exposed to SDK consumers

wzieba committed:
Update Experimentation api contract dump

wzieba committed:
Change `ExPlat#invalidateCache` signature to not return anything

wzieba committed:
Merge pull request #245 from Automattic/experimentation_make_anon_id_mandatory

Make anon id mandatory. Simplify API, remove `refresh`
